pandas/tests/test_window.py

```diff
@@ -275,6 +275,17 @@ def test_how_compat(self):
                             getattr(s, t)(freq='D', **kwargs), op)(how=how)
                         assert_series_equal(result, expected)
 
+    # GH #12373 : rolling functions error on float32 data
+    def test_ensure_type_float64(self):
+        expected = pd.Series([np.nan, np.nan, 2, 3, 4])
+
```

Comment:
@jreback 
I doubt that it won't raise on `M8[ns],m8[ns],datetime64[ns, UTC],category` because we already try to cast them to float so that we cannot distinguish them from `float`.

See https://github.com/pydata/pandas/blob/master/pandas/core/window.py#L441, we call `_prep_values()` which try to cast the `b.values` to `float`.

Here is the code for `_prep_values()` start from https://github.com/pydata/pandas/blob/master/pandas/core/window.py#L147

```Python
def _prep_values(self, values=None, kill_inf=True, how=None):

    if values is None:
        values = getattr(self._selected_obj, 'values', self._selected_obj)

    # coerce dtypes as appropriate
    if com.is_float_dtype(values.dtype):
        pass
    elif com.is_integer_dtype(values.dtype):
        values = values.astype(float)
    elif com.is_timedelta64_dtype(values.dtype):
        values = values.view('i8').astype(float)
    else:
        try:
            values = values.astype(float)
        except (ValueError, TypeError):
            raise TypeError("cannot handle this type -> {0}"
                            "".format(values.dtype))

    if kill_inf:
        values = values.copy()
        values[np.isinf(values)] = np.NaN

    return values
```
Maybe just change the `values.astype(float)` to com._asure_float64() will be a better idea compare to call this just before we really apply rolling functions from algos.

Doing so will also eliminate potential errors in other subclass which inherited from `_Window`. For example,  class `EWM`.
Here is an example code which call `ewma()` on `float64` vs `float32` data. Will error using current code  base

```Python
import pandas as pd
import numpy as np
ts = pd.Series(np.random.randn(20), index=pd.date_range('1/1/2000', periods=20))
ts.dtype
# Output: dtype('float64')
pd.ewma(ts, span=5) # Success!
ts1 = pd.Series(np.random.randn(20), index=pd.date_range('1/1/2000', periods=20), dtype=np.float32)
pd.ewma(ts1, span=5) # Will raise error
```

will produce
```python
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-9-bf62c08fbdd7> in <module>()
----> 1 pd.ewma(ts1, span=5)

C:\D\Projects\Github\pandas\pandas\stats\moments.py in ewma(arg, com, span, half
life, min_periods, freq, adjust, how, ignore_na)
    336                          adjust=adjust,
    337                          how=how,
--> 338                          ignore_na=ignore_na)
    339
    340

C:\D\Projects\Github\pandas\pandas\stats\moments.py in ensure_compat(dispatch, n
ame, arg, func_kw, *args, **kwargs)
    236                       FutureWarning, stacklevel=3)
    237
--> 238     result = getattr(r, name)(*args, **kwds)
    239
    240     if is_ndarray:

C:\D\Projects\Github\pandas\pandas\core\window.py in mean(self, **kwargs)
   1145     def mean(self, **kwargs):
   1146         """exponential weighted moving average"""
-> 1147         return self._apply('ewma', **kwargs)
   1148
   1149     @Substitution(name='ewm')

C:\D\Projects\Github\pandas\pandas\core\window.py in _apply(self, func, how, **k
wargs)
   1137                                  int(self.ignore_na), int(self.min_perio
ds))
   1138
-> 1139             results.append(np.apply_along_axis(func, self.axis, values))

   1140
   1141         return self._wrap_results(results, blocks, obj)

C:\Miniconda3\envs\pandas_dev\lib\site-packages\numpy\lib\shape_base.py in apply
_along_axis(func1d, axis, arr, *args, **kwargs)
     89     outshape = asarray(arr.shape).take(indlist)
     90     i.put(indlist, ind)
---> 91     res = func1d(arr[tuple(i.tolist())], *args, **kwargs)
     92     #  if res is a number, then we have a smaller output array
     93     if isscalar(res):

C:\D\Projects\Github\pandas\pandas\core\window.py in func(arg)
   1135                 def func(arg):
   1136                     return cfunc(arg, self.com, int(self.adjust),
-> 1137                                  int(self.ignore_na), int(self.min_perio
ds))
   1138
   1139             results.append(np.apply_along_axis(func, self.axis, values))


C:\D\Projects\Github\pandas\pandas\algos.pyx in pandas.algos.ewma (pandas\algos.
c:29634)()
    999 # Exponentially weighted moving average
   1000
-> 1001 def ewma(ndarray[double_t] input, double_t com, int adjust, int ignore_n
a, int minp):
   1002     '''
   1003     Compute exponentially-weighted moving average using center-of-mass.

ValueError: Buffer dtype mismatch, expected 'double_t' but got 'float'
```

My current `git log | head`
```
$ git log | head
commit 49f99a6e6df3eb68d11d580afc4021caa58b7efd
Author: Jeff Reback <jeff@reback.net>
Date:   Thu Feb 18 19:55:48 2016 -0500

    TST/COMPAT: xarray updated to 0.7.1 in conda, adjust to_xarray for
                MultiIndex frames with Categoricals

commit 528108bba4104b939bcfe6923677ddacc916ff00
Author: Chris Carroux <carroux@users.noreply.github.com>
Date:   Thu Feb 18 15:42:20 2016 -0600
```